pkg/mysqlsh/mysqlsh.go

```diff
@@ -0,0 +1,173 @@
+	c.SetStdout(&outb)
+	c.SetStderr(&errb)
+
+	if err := c.Run(); err != nil {
+		sout := sensitiveRegexp.ReplaceAllString(outb.String(), ":*****@")
+		serr := sensitiveRegexp.ReplaceAllString(errb.String(), ":*****@")
+		return errors.Wrapf(err, "run %s, stdout: %s, stderr: %s", cmd, sout, serr)
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) ConfigureInstance(ctx context.Context, instance string) error {
+	cmd := fmt.Sprintf(
+		"dba.configureInstance('%s', {'interactive': false, 'clearReadOnly': true})",
+		instance,
+	)
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "configure instance")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) AddInstance(ctx context.Context, clusterName, instance, podIp string) error {
+	opts := struct {
+		Interactive    bool   `json:"interactive"`
+		RecoveryMethod string `json:"recoveryMethod"`
+		WaitRecovery   int    `json:"waitRecovery"`
+		IpAllowList    string `json:"ipAllowList"`
+	}{
+		Interactive:    false,
+		RecoveryMethod: "clone",
+		WaitRecovery:   0,
+		IpAllowList:    fmt.Sprintf("%s,%s/8", defaultIPAllowList, podIp),
+	}
+
+	o, err := json.Marshal(opts)
+	if err != nil {
+		return errors.Wrap(err, "marshal options")
+	}
+
+	cmd := fmt.Sprintf("dba.getCluster('%s').addInstance('%s', %s)", clusterName, instance, string(o))
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "add instance")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) RejoinInstance(ctx context.Context, clusterName, instance string) error {
+	cmd := fmt.Sprintf("dba.getCluster('%s').rejoinInstance('%s', {'interactive': false})", clusterName, instance)
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "rejoin instance")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) RemoveInstance(ctx context.Context, clusterName, instance string) error {
+	cmd := fmt.Sprintf("dba.getCluster('%s').removeInstance('%s', {'interactive': false, 'force': true})", clusterName, instance)
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "remove instance")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) CreateCluster(ctx context.Context, clusterName, podIp string) error {
+	cmd := fmt.Sprintf("dba.createCluster('%s', {'ipAllowList': '%s,%s/8'})", clusterName, defaultIPAllowList, podIp)
+
+	if err := m.run(ctx, cmd); err != nil {
+		return errors.Wrap(err, "create cluster")
+	}
+
+	return nil
+}
+
+func (m *mysqlsh) DoesClusterExist(ctx context.Context, clusterName string) bool {
+	cmd := fmt.Sprintf("dba.getCluster('%s').status()", clusterName)
+	err := m.run(ctx, cmd)
+	return err == nil
+}
+
+func (m *mysqlsh) ClusterStatus(ctx context.Context, clusterName string) (innodbcluster.Status, error) {
+	var errb, outb bytes.Buffer
+
+	args := []string{"--result-format", "json", "--uri", m.uri, "--cluster", "--", "cluster", "status"}
+
+	c := m.exec.CommandContext(ctx, "mysqlsh", args...)
+	c.SetStdout(&outb)
+	c.SetStderr(&errb)
+
+	status := innodbcluster.Status{}
+
+	if err := c.Run(); err != nil {
```

Comment:
I think we can use a `m.run` method here if we change it to `func (m *mysqlsh) run(ctx context.Context, cmd string, ...args string) error`

Reply:
the reason of duplication here is not `args` but the unmarshaling the command output. I agree it's not optimal, but I intend to keep it as it is right now. What do you think?

Reply:
Missed the unmarshalling part. I think it's ok